File: hts_classifier/services/vector_store.py
from pathlib import Path

import chromadb

from ..core.config import settings
from ..data.processor import HTSEntry


# Collection names used during ingest and at query time
COLLECTION_AVG = "hts_entries"
COLLECTION_LEAF = "hts_entries_leaf"
COLLECTION_PATH = "hts_entries_path"


class VectorStore:
    def __init__(self, collection_name: str = COLLECTION_AVG) -> None:
        Path(settings.chroma_path).mkdir(parents=True, exist_ok=True)
        self._client = chromadb.PersistentClient(path=settings.chroma_path)
        self._collection = self._client.get_or_create_collection(
            name=collection_name,
            metadata={"hnsw:space": "cosine"},
        )

    @property
    def count(self) -> int:
        return self._collection.count()

    def get_all_ids(self) -> list[str]:
        return self._collection.get(include=[])["ids"]

    def upsert(self, entries: list[HTSEntry], embeddings: list[list[float]]) -> None:
        self._collection.upsert(
            ids=[e.hts_code for e in entries],
            embeddings=embeddings,  # ty: ignore[invalid-argument-type]
            documents=[e.path_string for e in entries],
            metadatas=[
                {
                    "hts_code": e.hts_code,
                    "description": e.description,
                    "path": " | ".join(e.path),
                    "indent": e.indent,
                    "general_rate": e.general_rate,
                }
                for e in entries
            ],
        )

    def query(self, embedding: list[float], top_k: int = 5) -> list[dict]:
        results = self._collection.query(
            query_embeddings=[embedding],
            n_results=top_k,
            include=["metadatas", "distances"],
        )
        assert results["metadatas"] is not None and results["distances"] is not None
        items = []
        for meta, dist in zip(results["metadatas"][0], results["distances"][0]):
            items.append(
                {
                    "hts_code": meta["hts_code"],
                    "description": meta["description"],
                    "path": meta["path"].split(" | "),
                    "indent": meta["indent"],
                    "general_rate": meta.get("general_rate", ""),
                    "score": 1.0 - dist,  # cosine distance -> similarity
                }
            )
        return items

File: src/services/vectorStore.ts
import { ChromaClient, IncludeEnum, type Collection, type IEmbeddingFunction } from "chromadb";
import { config } from "../config.js";

// Dummy embedding function — we always supply our own embeddings.
// ChromaDB v1 requires one to be set even when using precomputed embeddings.
const noopEmbedding: IEmbeddingFunction = {
  generate: async (texts: string[]) => texts.map(() => []),
};

export const COLLECTION_AVG = "hts_entries";
export const COLLECTION_LEAF = "hts_entries_leaf";
export const COLLECTION_PATH = "hts_entries_path";

export interface VectorResult {
  hts_code: string;
  description: string;
  path: string[];
  indent: number;
  general_rate: string;
  score: number;
}

export class VectorStore {
  private collection: Collection | null = null;
  private readonly name: string;
  private client: ChromaClient;

  constructor(collectionName: string = COLLECTION_AVG) {
    this.name = collectionName;
    this.client = new ChromaClient({ path: config.chromaUrl });
  }

  async init(): Promise<void> {
    this.collection = await this.client.getCollection({ name: this.name, embeddingFunction: noopEmbedding });
  }

  get count(): Promise<number> {
    return this.getCollection().count();
  }

  private getCollection(): Collection {
    if (!this.collection) throw new Error(`VectorStore '${this.name}' not initialized`);
    return this.collection;
  }

  async query(embedding: number[], topK: number = 5): Promise<VectorResult[]> {
    const col = this.getCollection();
    const results = await col.query({
      queryEmbeddings: [embedding],
      nResults: topK,
      include: [IncludeEnum.Metadatas, IncludeEnum.Distances],
    });

    const metadatas = results.metadatas?.[0] ?? [];
    const distances = results.distances?.[0] ?? [];

    return metadatas.map((meta, i) => ({
      hts_code: String(meta?.hts_code ?? ""),
      description: String(meta?.description ?? ""),
      path: String(meta?.path ?? "").split(" | "),
      indent: Number(meta?.indent ?? 0),
      general_rate: String(meta?.general_rate ?? ""),
      score: 1.0 - (distances[i] ?? 0), // cosine distance → similarity
    }));
  }
}
